### PR TITLE
Fix "Implement Members" action: prefer PSI to `Ty`s when generating PSI

### LINKED ISSUE #2982: `Implement members` intention doesn't handle where-blocks properly
## Environment

* **Intellij-Rust plugin version:** 0.2.0.2447-183-nightly
* **Rust toolchain version:** 1.31.0-nightly
* **IDE name and version:** IntelliJ IDEA Community 2018.3 EAP
* **Operating system:** macOS Mojave 10.14

## Problem description

The `Implement members` intention doesn't perform substitution of generic parameters in `where` blocks.  For example, it transforms this code
```rust
struct S;
trait T<'a, A> { fn  foo<P>(_: P) where P: T<'a, A>; }
impl <'b, B> T<'b, B> for S {}
```
into this
```rust
struct S;
trait T<'a, A> { fn  foo<P>(_: P) where P: T<'a, A>; }
impl <'b, B> T<'b, B> for S {
    fn foo<P>(_: P) where P: T<'a, A> {
        unimplemented!()
    }
}
```
(`'a` and `A` are not substituted in `where` block).

## Steps to reproduce

* Copy-paste code from problem description.

* Run `Implement members` intention.

### LINKED ISSUE #4967: "Implement Methods" expands primitive type aliases
## Environment

* **IntelliJ Rust plugin version:** 0.2.117.SNAPSHOT-201-dev
* **Rust toolchain version:** 1.41.0 (5e1a79984 2020-01-27) x86_64-apple-darwin
* **IDE name and version:** CLion 2020.1 EAP (CL-201.4865.10)
* **Operating system:** macOS 10.15.3

## Problem description

Unfortunately, #4938 does not fix #3082 as primitive type aliases are still expanded.

## Steps to reproduce

```rust
pub struct R;
pub type U = i32;
pub trait T {
    fn f() -> (R, U);
}
struct S;
impl T for S {}
``` 

Invoke "Implement Methods"

```rust
pub struct R;
pub type U = i32;
pub trait T {
    fn f() -> (R, U);
}
struct S;
impl T for S {
    fn f() -> (R, i32) {
        unimplemented!()
    }
}
```

### LINKED ISSUE #4979: Wrong type is imported when generating trait implementation
## Environment

* **IntelliJ Rust plugin version:** 0.2.117.SNAPSHOT-193-dev (#680fb11)
* **Rust toolchain version:** 1.41.0 (5e1a79984 2020-01-27) x86_64-unknown-linux-gnu
* **IDE name and version:** CLion 2019.3.3 (CL-193.6015.37)
* **Operating system:** Linux 5.0.0-37-generic

## Problem description
When you generate a method implementation for the `fmt::Display` trait, it imports a wrong type and the resulting code does not compile.

## Steps to reproduce
```rust
use std::fmt;

struct S;
impl fmt::Display for S {
    /*caret*/
}
```
generates this code after `ImplementMembersFix` is run:
```rust
use std::fmt;
use std::fmt::{Formatter, Error};

struct S;
impl fmt::Display for S {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        unimplemented!()
    }
}
```
which does not compile. `Error` is imported instead of `Result`.

### LINKED ISSUE #5044: "Generator Constructor" doesn't use type alias
## Environment

* **IntelliJ Rust plugin version:** 0.2.118.SNAPSHOT-201-dev
* **Rust toolchain version:** 1.41.0 (5e1a79984 2020-01-27) x86_64-apple-darwin
* **IDE name and version:** CLion 2020.1 EAP (CL-201.5616.31)
* **Operating system:** macOS 10.15.3

## Problem description

"Generator Constructor" action always uses original type instead of the alias. I guess it is #4944 regression.

## Steps to reproduce

```rust
type Coordinates = (u32, u32);

struct System {
    point: Coordinates,
}
```

call "Generator Constructor" on `System`

```rust
impl System {
    pub fn new(point: (u32, u32)) -> Self {
        System { point }
    }
}
```

expected

```rust
impl System {
    pub fn new(point: Coordinates) -> Self {
        System { point }
    }
}
```

### LINKED ISSUE #5592: Generic fn implementation ignores aliases
## Environment

* **IntelliJ Rust plugin version:** 0.2.124.3162-201
* **Rust toolchain version:** 1.45.0-nightly
* **IDE name and version:** CLion 2020.1
* **Operating system:** Windows 10

## Problem description
Generated fn implementations for generic traits ignore generic argument aliases and use unalias names.

## Steps to reproduce

```rust
use something::Type as AnotherType;

struct Widget;

impl From<AnotherType> for Widget {
    // generated impl uses Type istead of AnotherType
    fn from(_: Type) -> Self {
    }
}
```

### LINKED ISSUE #6085: Invalid type used when implementing functions
## Environment

* **IntelliJ Rust plugin version:** 0.3.130.3337-202
* **Rust toolchain version:** 1.48
* **IDE name and version:** CLion 2020.2
* **Operating system:** Windows 10

## Problem description

When implementing missing trait functions, generic type aliases are ignored and non-aliased type is used.

## Steps to reproduce

```
use foo::Bar as FooBar;

struct Widget;

impl From<FooBar> for Widget {

    // use autogenerated method implementation here and the following will appear:
    fn from(_: Bar) -> Self {
    }
}
```

### LINKED ISSUE #6340: Not all types are imported by Implement Members quick fix
## Environment

* **IntelliJ Rust plugin version:** 0.3.135.3449-202-nightly
* **Rust toolchain version:** 1.49.0-nightly (ffa2e7ae8 2020-10-24) x86_64-apple-darwin
* **IDE name and version:** IntelliJ IDEA 2020.2.3 Ultimate Edition (IU-202.7660.26)
* **Operating system:** macOS 10.15.6

## Problem description
When I try to invoke `Implement members` quick fix for `Display` trait, it doesn't import `std::fmt::Result` type alias that leads to error annotation.

## Steps to reproduce
```rust
struct Foo;

impl std::fmt::Display for Foo {
}
```

Expected:
```rust
use std::fmt::{Formatter, Result};

struct Foo;

impl std::fmt::Display for Foo {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        unimplemented!()
    }
}
```

Actual:
```rust
use std::fmt::Formatter;

struct Foo;

impl std::fmt::Display for Foo {
    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
        unimplemented!()
    }
}
```

### PR DESCRIPTION
**Pull request by vlad20012**
Fixes #4967, fixes #5044, fixes #4979, fixes #6340, fixes #2982, fixes #5592, fixes #6085

1. Never expand type aliases (#4967, #5044)
2. Don't expand macro types
3. Auto-import trait names from type bounds (#2982)
4. Resolve name conflicts during auto-import (#4979, #6340):
   1. If the target item is in the scope with a different name (`use A as B`), use that name (#5592, #6085)
   2. If the name exists in the scope, but refers to another item, use a qualified path to the target item
   3. If the target can't be imported, use crate-qualified path to the item (produces incorrect code, but it should be simpler to fix)

changelog:

### CODE REVIEW

**Comment by Kobzol** [comment]
This is awesome! :rocket: I know that this started with type aliases, but point 4) is actually maybe even better :laughing:

This also fixes https://github.com/intellij-rust/intellij-rust/issues/6340, which is basically a duplicate of https://github.com/intellij-rust/intellij-rust/issues/4979.

I found a few more cases where the type was rendered instead of the type reference being used. I changed them [here](https://github.com/Kobzol/intellij-rust/commit/a214d3d8dc8535901212afdf9721ca7c3f929443). Basically uses of `renderInsertionSafe` are now potentially suspicious :)

BTW, when I run CLion with this PR, I always get `java.lang.RuntimeException: Broken stub format, most likely version of FILE (Rust.file) was not updated after serialization changes` on startup.

**Comment by Kobzol** [comment]
Btw, would it still make sense after this PR to move `aliasedBy` from `TyAdt` to `Ty` to better support scenarios where we only have a type from type inference?

**Comment by vlad20012** [comment]
> BTW, when I run CLion with this PR, I always get `java.lang.RuntimeException: Broken stub format, most likely version of FILE (Rust.file) was not updated after serialization changes` on startup.

Hmm, I can't reproduce. Maybe previously you were on some branch with `STUB_VERSION = 215`? 

> I found a few more cases where the type was rendered instead of the type reference being used. I changed them [here](https://github.com/Kobzol/intellij-rust/commit/a214d3d8dc8535901212afdf9721ca7c3f929443). Basically uses of `renderInsertionSafe` are now potentially suspicious :)

:+1: but usage of `substAndGetText` with `Substitution` is also suspicious. Aliases from a `Substitution` are still expanded, for example in
```rust
type Alias = i32;
struct Bar<T> { f: T }
impl Bar<Alias> {}
```

I've introduced `PsiSubstitution` to solve this. You can obtain it from a path with `pathPsiSubst()` function (in the case of impl for struct, the path is `impl.typeReference?.skipParens() as? RsBaseType)?.path`)

> Btw, would it still make sense after this PR to move `aliasedBy` from `TyAdt` to `Ty` to better support scenarios where we only have a type from type inference?

Yes!

**Comment by Kobzol** [comment]
>Hmm, I can't reproduce. Maybe previously you were on some branch with STUB_VERSION = 215?

Indeed I was. Probably a false alarm.

> but usage of substAndGetText with Substitution is also suspicious. Aliases from a Substitution are still expanded, for example in

I see. Will you include the changes in getter/setter in this PR or do you want me to open another PR on top of this one where getter/setter generation is fixed?

**Review by Kobzol — APPROVED** [review]

File: src/main/kotlin/org/rust/ide/refactoring/implementMembers/impl.kt [hunk 1]
```diff
@@ -154,82 +180,39 @@ private fun createExtraWhitespacesAroundFunction(left: PsiElement, right: PsiEle
...
+            when (it) {
+                is RsConstant -> {
+                    val initialValue = RsDefaultValueBuilder(it.knownItems, it.containingMod, factory, true)
+                        .buildFor(it.typeReference?.type?.substitute(subst) ?: TyUnknown, emptyMap())
+                    "    const ${it.nameLikeElement.text}: ${it.typeReference?.renderTypeReference() ?: "_"} = ${initialValue.text};"
+                }
+                is RsTypeAlias ->
+                    "    type ${it.escapedName} = ();"
+                is RsFunction ->
+                    "    ${renderFunctionSignature(it) ?: ""}{\n        todo!()\n    }"
```
Comment (mchernyavsky) [inline src/main/kotlin/org/rust/ide/refactoring/implementMembers/impl.kt:203]: `?: ""` is not needed.

File: src/main/kotlin/org/rust/ide/refactoring/implementMembers/impl.kt [hunk 2]
```diff
@@ -154,82 +180,39 @@ private fun createExtraWhitespacesAroundFunction(left: PsiElement, right: PsiEle
...
-    val subst = trait.subst
-    val body = members.joinToString(separator = "\n", transform = {
-        when (it) {
-            is RsConstant -> {
-                val initialValue = RsDefaultValueBuilder(it.knownItems, it.containingMod, this, true)
-                    .buildFor(it.typeReference?.type?.substitute(subst) ?: TyUnknown, emptyMap())
-                "    const ${it.nameLikeElement.text}: ${it.typeReference?.substAndGetText(subst) ?: "_"} = ${initialValue.text};"
+private class MembersGenerator(
+    private val factory: RsPsiFactory,
+    private val impl: RsImplItem,
```
Comment (mchernyavsky) [inline src/main/kotlin/org/rust/ide/refactoring/implementMembers/impl.kt:185]: `val` is not needed.

File: src/main/kotlin/org/rust/lang/core/psi/ext/RsFnPointerType.kt [hunk 3]
```diff
@@ -6,7 +6,27 @@
 package org.rust.lang.core.psi.ext
 
 import org.rust.lang.core.psi.RsFnPointerType
+import org.rust.lang.core.psi.RsFunction
```
Comment (mchernyavsky) [inline src/main/kotlin/org/rust/lang/core/psi/ext/RsFnPointerType.kt:9]: Unused

File: src/main/kotlin/org/rust/ide/presentation/RsPsiRenderer.kt [hunk 4]
```diff
@@ -0,0 +1,715 @@
...
+ * found in the LICENSE file.
+ */
+
+package org.rust.ide.presentation
+
+import org.jetbrains.annotations.Nullable
+import org.rust.ide.utils.import.ImportCandidate
+import org.rust.ide.utils.import.ImportCandidatesCollector
+import org.rust.ide.utils.import.ImportContext
+import org.rust.ide.utils.import.RsImportHelper
```
Comment (mchernyavsky) [inline src/main/kotlin/org/rust/ide/presentation/RsPsiRenderer.kt:12]: Unused

File: src/main/kotlin/org/rust/ide/presentation/RsPsiRenderer.kt [hunk 5]
```diff
@@ -0,0 +1,715 @@
...
+    selfParameter: RsSelfParameter,
+    renderLifetimes: Boolean
+): String = buildString { appendSelfParameter(this, selfParameter, renderLifetimes) }
+
+fun RsPsiRenderer.renderFunctionSignature(
+    fn: RsFunction,
+    renderLifetimes: Boolean
+): String = buildString { appendFunctionSignature(this, fn, renderLifetimes) }
+
+data class PsiRenderingOptions(
```
Comment (mchernyavsky) [inline src/main/kotlin/org/rust/ide/presentation/RsPsiRenderer.kt:69]: Unused

**Review by mchernyavsky — APPROVED** [review]
LGTM except for minor comments

**Comment by vlad20012** [comment]
> Will you include the changes in getter/setter in this PR or do you want me to open another PR on top of this one where getter/setter generation is fixed?

I'd prefer you. If it doesn't bother you...

**Comment by vlad20012** [comment]
:rocket: :rocket: :rocket: 
bors r+

**Comment by Kobzol** [comment]
> I'd prefer you. If it doesn't bother you...

No problem, I'll send it right after this PR gets merged :)

**Pull request merged**